contracts/messaging/gateway/FuelERC20Gateway.sol

```diff
@@ -95,31 +98,59 @@ contract FuelERC20Gateway is
         return _deposits[tokenAddress][fuelTokenId];
     }
 
-    /// @notice Deposits the given tokens to an address on Fuel
-    /// @param to Fuel account or contract to deposit tokens to
+    /// @notice Deposits the given tokens to an account on Fuel
+    /// @param to Fuel account to deposit tokens to
     /// @param tokenId ID of the token being transferred to Fuel
     /// @param fuelTokenId ID of the token on Fuel that represent the deposited tokens
     /// @param amount Amount of tokens to deposit
     /// @dev Made payable to reduce gas costs
     function deposit(bytes32 to, address tokenId, bytes32 fuelTokenId, uint256 amount) external payable whenNotPaused {
-        require(amount > 0, "Cannot deposit zero");
-
-        //transfer tokens to this contract and update deposit balance
-        IERC20Upgradeable(tokenId).safeTransferFrom(msg.sender, address(this), amount);
-        _deposits[tokenId][fuelTokenId] = _deposits[tokenId][fuelTokenId] + amount;
-
-        //send message to gateway on Fuel to finalize the deposit
-        bytes memory data = abi.encodePacked(
+        bytes memory messageData = abi.encodePacked(
             fuelTokenId,
             bytes32(uint256(uint160(tokenId))),
             bytes32(uint256(uint160(msg.sender))), //from
             to,
             bytes32(amount)
         );
-        sendMessage(CommonPredicates.CONTRACT_MESSAGE_PREDICATE, data);
+        _deposit(tokenId, fuelTokenId, amount, messageData);
+    }
 
-        //emit event for successful token deposit
-        emit Deposit(bytes32(uint256(uint160(msg.sender))), tokenId, fuelTokenId, amount);
+    /// @notice Deposits the given tokens to a contract on Fuel with optional data
+    /// @param to Fuel account or contract to deposit tokens to
+    /// @param tokenId ID of the token being transferred to Fuel
+    /// @param fuelTokenId ID of the token on Fuel that represent the deposited tokens
+    /// @param amount Amount of tokens to deposit
+    /// @param data Optional data to send with the deposit
+    /// @dev Made payable to reduce gas costs
+    function depositWithData(
```

Comment:
So in this case, the flow would be to call process_message on the BridgeFungibleToken contract, ehich wouod mint and send tokens to the specified fuel contract's process message function? or would the BridgeFungibleToken be bypassed ?

Reply:
I think what I'm unsure of atm is:
Is this mechanism strictly for bridging tokens directly to a contract on Fuel, and optionally executing code on that contract, or is this meant to be more general-purpose, allowing a message from Ethereum to trigger execution of code an any fuel contract?

Reply:
This is a function on the FuelERC20Gateway contract, so the message being generated is always going to be related to bridging ERC20s. The added bit is to be able to call `process_message` and send these ERC20 tokens on another contract that will care about the extra data attached. It will be seeing a message rooted in the ERC20 bridging process, but this is a feature as some of the data (particularly the "from" field) could be important to know.